x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java

```diff
@@ -73,14 +106,20 @@ private synchronized void processResult(PyTorchResult result) {
         }
     }
 
-    public PyTorchResult waitForResult(String requestId, TimeValue timeout) throws InterruptedException {
-        PendingResult pendingResult = pendingResults.computeIfAbsent(requestId, k -> new PendingResult());
-        try {
-            if (pendingResult.latch.await(timeout.millis(), TimeUnit.MILLISECONDS)) {
-                return pendingResult.result;
-            }
-        } finally {
-            pendingResults.remove(requestId);
+    public PyTorchResult waitForResult(
+        NativePyTorchProcess process,
+        String requestId,
+        PendingResult pendingResult,
+        TimeValue timeout
+    ) throws InterruptedException {
+        if (process == null || process.isProcessAliveAfterWaiting() == false) {
```

Comment:
It would be nice to avoid the call to `isProcessAliveAfterWaiting()` as it contains a 45ms wait. 

It be sufficient to know that `while` loop in `process()` is still running by setting a boolean when the loop exits. That boolean would be tested here and if the `while` has finished we know there is no point waiting for a result that will never come.